### PR TITLE
Bids 1770/explorer configuration administration

### PR DESCRIPTION
**Pull request by MauserBitfly**

### CODE REVIEW

**Review by D13ce — CHANGES_REQUESTED** [review]

File: tables.sql [hunk 1]
```diff
@@ -1057,8 +1057,7 @@ create table node_jobs
     primary key (id)
 );
 
-drop table if exists ad_configurations;
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply tables.sql:1060]: based on a comment by @recy21 (which I cannot find anymore) the drop table was replaced with `if not exists` in the create table statement.

File: templates/user/explorer_configuration.html [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      <form action="/user/explorer_configuration" id="explorerConfigurationForm" method="POST">
+        {{ .CsrfField }}
+        {{ range $category, $keyMap := .Configurations }}
+          <div class="card p-3">
+            <h3 class="pb-3">{{ $category }}</h3>
+            {{ range $key, $configValue :=  $keyMap }}
+              <div>
+                {{ if eq $configValue.DataType "int" }}
+                  <input type="number" name="{{ $category }}-{{ $key }}" value="{{ $configValue.Value }}" />
+                {{ else if eq $configValue.DataType "bool" }}
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply templates/user/explorer_configuration.html:18]: @D13ce : There is no restriction on the data types in the backend ... the line you refer to is just from a helper method casting int data types.  The UI for the bool I just thought might come in handy if we need it.

File: types/templates.go [hunk 3]
```diff
@@ -205,10 +206,9 @@ type IndexPageData struct {
 }
 
 type SlotVizPageData struct {
-	Epochs        []*SlotVizEpochs
-	Selector      string
-	HardforkEpoch uint64
-	HardforkName  string
+	Epochs   []*SlotVizEpochs
+	Selector string
+	Config   ExplorerConfigurationKeyMap
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply types/templates.go:211]: The Idea here is that if we need an additional config for the Frontend we just need to add it to the DefaultExplorerConfiguration without any mapping in the go code.

File: types/templates.go [hunk 4]
```diff
@@ -1923,6 +1928,49 @@ type AdConfig struct {
...
+	ExplorerConfigValue
+}
+type ExplorerConfigurationKeyMap map[ExplorerConfigurationKey]ExplorerConfigValue
+type ExplorerConfigurationMap map[ExplorerConfigurationCategory]ExplorerConfigurationKeyMap
+
+func (configMap ExplorerConfigurationMap) GetConfigValue(category ExplorerConfigurationCategory, configKey ExplorerConfigurationKey) (ExplorerConfigValue, error) {
+	configValue := ExplorerConfigValue{}
+	kyMap, ok := configMap[category]
+	if ok {
+		configValue, ok = kyMap[configKey]
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply types/templates.go]: fixed

File: handlers/index.go [hunk 5]
```diff
@@ -45,12 +45,19 @@ func Index(w http.ResponseWriter, r *http.Request) {
...
+			if err == nil && visibleTo >= data.CurrentEpoch {
+				pageData.SlotVizData =
+					types.SlotVizPageData{
+						Epochs:   services.LatestSlotVizMetrics(),
+						Selector: "slotsViz",
+						Config:   configuration[services.ConfigurationCategorySlotViz],
+					}
+			}
 		}
 	}
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply handlers/index.go]: I have put it in a separate function, but left it in the index.go file as the logic is specific to the index page. And separated the functions a bit so you can read it better. ;-) 
As I don't think loading the config is vital I didn't log the error messages.

File: services/configuration.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+	result, err := db.GetExplorerConfigurations()
+
+	if err != nil {
+		return nil, err
+	}
+
+	configs := types.ExplorerConfigurationMap{}
+
+	// first we clone de defaults
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply services/configuration.go]: As "de" to cool for you, I change it "the"

File: types/templates.go [hunk 7]
```diff
@@ -1923,6 +1928,49 @@ type AdConfig struct {
...
+		if ok {
+			return configValue, nil
+		}
+	}
+	return configValue, errors.New("error")
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply types/templates.go]: fixed

File: types/templates.go [hunk 8]
```diff
@@ -1923,6 +1928,49 @@ type AdConfig struct {
...
+		if configValue.DataType != "int" {
+			return 0, errors.New("Wrong data type")
+		} else {
+			return strconv.ParseUint(configValue.Value, 10, 64)
+		}
+	}
+	return 0, err
+}
+
+func (configMap ExplorerConfigurationMap) GetValue(category ExplorerConfigurationCategory, configKey ExplorerConfigurationKey) (string, error) {
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply types/templates.go]: renamed

File: templates/user/explorer_configuration.html [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+                  <input type="checkbox" name="{{ $category }}-{{ $key }}" {{ if not (len .Config.Enabled 0) }}checked{{ end }} />
+                {{ else }}
+                  <input type="string" name="{{ $category }}-{{ $key }}" value="{{ $configValue.Value }}" />
+                {{ end }}
+                <label class="ml-3" for="{{ $category }}-{{ $key }}">{{ $key }}</label>
+              </div>
+            {{ end }}
+          </div>
+        {{ end }}
+        <button type="submit" class="btn btn-primary w-100">Save</button>
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply templates/user/explorer_configuration.html]: added a margin - hope it now is usable for you ;-)

File: services/configuration.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (D13ce) [reply services/configuration.go]: Thank you, youngling.

File: templates/user/explorer_configuration.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (D13ce) [reply templates/user/explorer_configuration.html:18]: The line link I have chosen wasn't the best, sorry.

My point is the following: We do have a `GetUInt64Value` and a `GetStringValue` but no `GetBooleanValue`. I agree with you, Booleans should be supported. But they should be supported completely and that includes a helper function for Booleans as well. Thoughts?

[the remark this answers is not in the document]
Comment (D13ce) [reply templates/user/explorer_configuration.html:18]: As discussed, adding it to the frontend was the harder part as the input field differs. Adding it to the backend (i.e. implementing the helper function) is super simple and will be done on demand.

File: types/templates.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (D13ce) [reply types/templates.go:211]: I am sorry, that was my bad. I misread the code and thought that the WHOLE explorer configuration is part of `SlotVizPageData`. That is not the case though, only the keymap for `ConfigurationCategorySlotViz` is added (and yes, CURRENTLY this is the whole configuration, but you get what I mean).

File: handlers/index.go [hunk 10]
```diff
@@ -80,3 +66,26 @@ func IndexPageData(w http.ResponseWriter, r *http.Request) {
...
+
+func getSlotVizData(currentEpoch uint64) types.SlotVizPageData {
+	var slotVizPageData types.SlotVizPageData
+
+	var visiblFrom uint64
+	var visibleTo uint64
+	configuration, err := services.GetExplorerConfigurationsWithDefaults()
+	if err == nil {
+		visiblFrom, err = configuration.GetUInt64Value(services.ConfigurationCategorySlotViz, services.ConfigurationKeyVisibleFromEpoch)
+	}
```
Comment (D13ce) [inline handlers/index.go:78]: **Minor:** I would prefer early exits when errors appear like this:
```suggestion
	configuration, err := services.GetExplorerConfigurationsWithDefaults()
	if err != nil {
		return slotVizPageData
	}
	
	visiblFrom, err = configuration.GetUInt64Value(services.ConfigurationCategorySlotViz, services.ConfigurationKeyVisibleFromEpoch)

```

File: handlers/index.go [hunk 11]
```diff
@@ -80,3 +66,26 @@ func IndexPageData(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 }
+
+func getSlotVizData(currentEpoch uint64) types.SlotVizPageData {
```
Comment (D13ce) [inline handlers/index.go:70]: This function does not have any error handling. If errors happen, the slot viz will simply not be shown but there is no information, WHAT went wrong.

This is a delicate topic. For example: We might not want to throw an error log if configuration is missing. But if something else goes wrong? What are your thoughts?

File: templates/user/explorer_configuration.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (D13ce) [reply templates/user/explorer_configuration.html]: LOVE it!

**Review by D13ce — COMMENTED** [review]
Left a question regarding error handling. Once that is answered, this PR can be merged.

**Review by D13ce — APPROVED** [review]
lgtm, nice feature!

**Pull request merged**